### DISCUSSION TITLE
Testing - code & hardware based

### DISCUSSION THREAD

**Discussion by OOHehir** [LIFsCode/ELOC-3.0#70] [General]
This discussion is in response to #59, the requirement for testing.

I'm fairly certain this will be an ongoing requirement & would benefit from automation so that the code is continuously tested in response to updates. I believe it will be very difficult to achieve development & testing at a reasonable pace without it.

I previously tried testing with the [PlatformIO feature](https://piolabs.com/blog/insights/cicd-testing-coverage-versioning.html) but found it difficult because of all the file dependencies. This made building the test code complicated. I'll revisit this.

I think testing with hardware-in-the-loop (HIL) would be very beneficial & will look at this also. 

Whatever I find I'll post here...

**Comment by OOHehir** [comment]
@EDsteve @LIFsCode I had some thoughts about the HIL testing. 

Long term the ideal scenario would be that a commit triggers a GitHub action to compile the code, push it to some device which flashes it to the ELOC, runs tests & post the results on Github. But that involves a lot of setup & time..

It would seem that most of the current bugs center around state changes (i.e. recording, Bluetooth & inference). Therefore what might offer the quickest payback in terms of time & effort is manually flashing an ELOC & connecting it to some device (perhaps another ESP32, RPi etc., lets call it the 'supervisor' ) that could monitor the ELOC. The supervisor would monitor the terminal output of the ELOC & repeatedly trigger state changes using spare GPIO pins on the ELOC (would require ELOC firmware modification). If the supervisor detects a fault/ restart it would commit the terminal output as a text file to the Github repo (say the previous 50 lines of terminal output). Perhaps longer term this could be achieved using Bluetooth from the supervisor to the ELOC.

The terminal output could be either parsed for keywords (e.g. Guru, brownout, the initial output of "------------------------- ELOC Recorder -------------------------\n" etc..) or another spare GPIO on the ELOC board could be toggled at startup to indicate that the device has restarted & hence the previous terminal output should be captured.

There's plenty of pain points here, trying to get the supervisor to commit a file to GitHub (needs authentication, more than a simple password), buffering the terminal output so that when a restart occurs it can be committed & toggling the GPIO pins at the appropriate times. Saying that it seems like it could be useful & relieve some of the time consuming effort of testing..

**Comment by EDsteve** [comment]
@OOHehir That sounds like the perfect testing environment indeed and would save a lot of time. 
Since we are a bit under time pressure to finish >20 ELOCs till 15th of February (Hope we have a stable running firmware by then). It will have to wait a bit to take a closer look into that solution. As you said. It will take a some effort to set that up.

For now i just let a few ELOCs run all the time, knock every few hours on the case and connect once a day with BT. The results i put in a spreadsheet and save the logs and configs for each test.

**Comment by OOHehir** [comment]
@LIFsCode Now that unit testing is enabled I was checking if you'd any thoughts on system testing?

The only approach I can see that makes automated testing seem feasible is using the serial terminal. An external test application could inject commands via the serial terminal & parse the serial output for the desired response.

It seems the components that sense/accept external input (e.g. BT, accelerometer, button, SD card) would need testing. A class could be created to accept the serial input & dispatch it to the relevant component. The component code would, as closely as possible, inject this command as if it had originated from the hardware.

Something like this together with unit testing could give significant coverage of the code.

I'm happy to develop the remote application & the class for handing the serial input.

**Reply by LIFsCode** [reply]
@OOHehir Good idea 👍 I think the serial terminal is the best way to do this.

Since the ELOC usually is controlled via BT it would be good to also have a remote application which can access Bluetooth and include that interface in an automatic test setup. 
Maybe a raspberry could be a suitable platform (cheap, having all the interfaces, IOs)?

**Comment by OOHehir** [comment]
@LIFsCode @EDsteve I've just [added](https://github.com/LIFsCode/ELOC-3.0/tree/master/eloc610LowPowerPartition/tools/external_tests/robot) some system testing for the BT serial interface. The README covers most of the detail. I went with an established framework rather than trying to reinvent the wheel. It should work cross platform (including from desktop), but you may need to edit the BT serial port for Windows. The tests are very simple at the moment but they should be straightforward to improve & extend.

I'll hopefully be able to use the same framework to add injection of commands from the serial terminal into the running application.
